### PR TITLE
Finance: Credit card fraud detection using FL pipeline

### PR DESCRIPTION
**Pull request by majercakdavid**
## Purpose
The example in this PR should primarily demonstrate capabilities of Federated Learning (FL) pipeline for use case in finance industry, in particular **credit card fraud detection**. Alongside this goal, it also shows how to setup different types of neural networks (NN) for FL scenarios, namely: **Dense ANN**, **LSTM**, **VAE**.

Dataset used: https://www.kaggle.com/datasets/kartik2112/fraud-detection?datasetId=817870&sortBy=voteCount&types=competitions
Dataset generated using following toolkit(MIT License): https://github.com/namebrandon/Sparkov_Data_Generation

## Does this introduce a breaking change?

```
[ ] Yes
[X] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
* Get the code
* Setup federated learning environment in AML
* Adjust parameters in `examples/pipelines/ccfraud/config.yaml` and also edit file names for your computes in `examples/components/CCFRAUD/preprocessing/config` folder (this is used also as an example on how to load config file based on the compute cluster name)
* Run pipeline in the `examples/pipelines/ccfraud` folder

**Successful runs**:

- **LSTM**: https://ml.azure.com/experiments/id/c175dc31-0e64-45df-8508-1d3fad7f332f/runs/plucky_battery_kjnv9zwy4j?wsid=/subscriptions/48bbc269-ce89-4f6f-9a12-c6f91fcb772d/resourcegroups/damajercak-fl-0/workspaces/aml-fldemodm0&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#
- **VAE**: https://ml.azure.com/runs/frosty_papaya_vdhmdwpd31?wsid=/subscriptions/48bbc269-ce89-4f6f-9a12-c6f91fcb772d/resourcegroups/damajercak-fl-0/workspaces/aml-fldemodm0&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#
- **Linear**: https://ml.azure.com/runs/blue_sun_djst2jsfmb?wsid=/subscriptions/48bbc269-ce89-4f6f-9a12-c6f91fcb772d/resourcegroups/damajercak-fl-0/workspaces/aml-fldemodm0&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#

### CODE REVIEW

File: examples/components/CCFRAUD/training/run.py [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+                # log metrics at the pipeline level
+                for name, value in train_metrics.get_global().items():
+                    log_message.append(f"{name}: {value}")
+                    self.log_metrics(
+                        mlflow_client,
+                        root_run_id,
+                        name,
+                        value,
+                        pipeline_level=True,
+                    )
```
Comment (garg-amit) [inline examples/components/CCFRAUD/training/run.py:296]: this metric seems to be a bit off. Can you pass `pipeline_level=False` if you are logging on each epoch? Also, let's log one more train and test metric for each FL iteration (`pipeline_level=True`). You can take reference from the MNIST example.

File: examples/components/CCFRAUD/training/datasets.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
+import torch
+from torch.utils.data import Dataset
+
+
```
[the remark this answers is not in the document]
Comment (majercakdavid) [reply examples/components/CCFRAUD/training/datasets.py:4]: I thought that it's a typo, isn't it? I believe it could be nice to rename them all, what do you think?

[the remark this answers is not in the document]
Comment (garg-amit) [reply examples/components/CCFRAUD/training/datasets.py:4]: oh that actually means "Train in silo"

File: examples/components/CCFRAUD/preprocessing/run.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    parser = get_arg_parser()
+    # run the parser on cli args
+    args = parser.parse_args(cli_args)
+    logger.info(f"Running script with arguments: {args}")
+
+    # Get runtime specific configuration
+    run: Run = Run.get_context()
+    compute_target = run.get_details()["target"]
+    logger.info(f"Compute target: {compute_target}")
+    config_name = (
```
[the remark this answers is not in the document]
Comment (garg-amit) [reply examples/components/CCFRAUD/preprocessing/run.py:231]: yup! if the file has only a region then you can set that param for each silo in the `examples/pipelines/ccfraud/config.yaml`. That might simplify the code as well as give users a bit more flexibility.

File: examples/components/CCFRAUD/training/run.py (hunk 1 above)
Comment (majercakdavid) [reply examples/components/CCFRAUD/training/run.py:296]: Hi @garg-amit, passing `pipeline_level=True` is in this case intentional as this should be the logging the metrics for whole epoch and I expected that this is the correct way. I believe that as a user you may want to have an overview of progress within nodes for each epoch at pipeline level. What are your thoughts on that?

File: examples/pipelines/ccfraud/config.yaml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  resource_group_name: "<RESOURCE_GROUP>"
+  workspace_name: "<AML_WORKSPACE_NAME>"
+
+# federated learning parameters
+federated_learning:
+  orchestrator:
+    compute: cpu-orchestrator
+    datastore: datastore_orchestrator
+
+  silos:
```
[the remark this answers is not in the document]
Comment (thomasp-ms) [reply examples/pipelines/ccfraud/config.yaml:15]: Nevermind, we talked and now I do understand that the silos are actually using different data. It is okay to leave as-is for this PR, but eventually we want to decouple the splitting job from the training job, since the graph just shows 2 datasets and that will be confusing to the users.

File: examples/components/CCFRAUD/preprocessing/run.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        None
+    """
+
+    logger.info(
+        f"Raw Training Data path: {raw_training_data}, Raw Testing Data path: {raw_testing_data}, Processed Training Data dir path: {train_data_dir}, Processed Testing Data dir path: {test_data_dir}"
+    )
+
+    logger.debug('Loading regions from "us_regions.csv"')
+    global STATES_REGIONS
+    df_states_regions = pd.read_csv(hydra.utils.get_original_cwd() + "/us_regions.csv")
```
[the remark this answers is not in the document]
Comment (majercakdavid) [reply examples/components/CCFRAUD/preprocessing/run.py:151]: Actually, for this one you have to use hydra because using hydra sets os cwd to its own output folder and thus in order to get original execution cwd one need to use hydra api

File: examples/components/CCFRAUD/training/run.py (hunk 1 above)
Comment (garg-amit) [reply examples/components/CCFRAUD/training/run.py]: Right, @majercakdavid! We definitely need to provide a way for users to look at metrics for each epoch at the pipeline level but all the metrics by default show the metrics at the pipeline level rn. What I meant with `pipeline_level=True` param is to report metrics for each iteration (along with metrics for each epoch). I know the param name is a bit confusing- it should be named as `iteration_level`. Happy to elaborate on it on Teams if it's still confusing!

File: examples/pipelines/ccfraud/submit.py [hunk 6]
```diff
@@ -0,0 +1,357 @@
+"""Federated Learning Cross-Silo basic pipeline.
```
[the remark this answers is not in the document]
Comment (ShizeSu06) [reply examples/pipelines/ccfraud/submit.py:1]: +1

File: docs/real-world-examples/ccfraud.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
+# Finance example for Federated Learning in Azure ML
```
[the remark this answers is not in the document]
Comment (ShizeSu06) [reply docs/real-world-examples/ccfraud.md:1]: I saw David implemented those different type NNs in the models.py at traininsilo component, but I agree with Thomas that, in addition to the code there, it might be valuable to add a very brief mention about that in the doc here, and also maybe added some minor extra comments/explanations in the models.py file (which currently mainly just codes), adding 1-2 sentence there for each type NN in models.py might probably good enough.

[the remark this answers is not in the document]
Comment (majercakdavid) [reply docs/real-world-examples/ccfraud.md:1]: @ShizeSu06 - added in this file at very bottom :)

**Comment by majercakdavid** [comment]
**Data splitting submission:** https://ml.azure.com/experiments/id/5a079ea9-aca8-4a4b-8899-fedfe49069af/runs/sincere_glove_wxllws5538?wsid=/subscriptions/48bbc269-ce89-4f6f-9a12-c6f91fcb772d/resourcegroups/damajercak-fl-0/workspaces/aml-fldemodm0&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#

**Training using CPU computes:** https://ml.azure.com/experiments/id/c175dc31-0e64-45df-8508-1d3fad7f332f/runs/elated_owl_4h85p705cb?wsid=/subscriptions/48bbc269-ce89-4f6f-9a12-c6f91fcb772d/resourcegroups/damajercak-fl-0/workspaces/aml-fldemodm0&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#

File: docs/real-world-examples/ccfraud.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+Federated Learning (FL) is an approach where we train a single ML models on a distinct datasets, usually in distinct location which cannot be gathered in one central location for reasons like compliance, security or many others. This is more concern in financial field than most probably in any other as this data contain highly sensitive information about either people or institutions.
+
+Therefore, we came up with and example how an institutions like a banks or credit card companies can train a model for fraud detection in federating learning fashion.
+
+:warning: Please familiarize yourself with basics and remarks mentioned in the [**documentation**](../../../README.md) file in the root of the repository.
+
+:warning: This is an example for demonstration purposes only and the authors of this repository do not take any responsibility for loss or harm made by its usage
+
+# Data
```
[the remark this answers is not in the document]
Comment (majercakdavid) [reply docs/real-world-examples/ccfraud.md:12]: done :)

File: examples/components/CCFRAUD/preprocessing/run.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (majercakdavid) [reply examples/components/CCFRAUD/preprocessing/run.py]: No need to load the config anymore - resolving

File: examples/components/CCFRAUD/training/run.py (hunk 1 above)
Comment (majercakdavid) [reply examples/components/CCFRAUD/training/run.py]: Hi @garg-amit please can you confirm that the updated solution is what you expected and resolve the conversation? Thank you :)

File: examples/components/CCFRAUD/preprocessing/conda.yaml [hunk 9]
```diff
@@ -0,0 +1,16 @@
+name: finance_preprocess_conda_env
```
Comment (garg-amit) [inline examples/components/CCFRAUD/preprocessing/conda.yaml:1]: how about renaming this to `ccfraud_preprocess_conda_env`?

File: examples/components/CCFRAUD/preprocessing/preprocessing.yaml [hunk 10]
```diff
@@ -0,0 +1,39 @@
+
```
Comment (garg-amit) [inline examples/components/CCFRAUD/preprocessing/preprocessing.yaml:1]: is this file required?

File: examples/components/CCFRAUD/preprocessing/run.py [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+    # build an arg parser
+    parser = get_arg_parser()
+    # run the parser on cli args
+    args = parser.parse_args(cli_args)
+    logger.info(f"Running script with arguments: {args}")
+
+    # Get runtime specific configuration
+    run: Run = Run.get_context()
+    compute_target = run.get_details()["target"]
+    logger.info(f"Compute target: {compute_target}")
```
Comment (garg-amit) [inline examples/components/CCFRAUD/preprocessing/run.py:212]: why do we need a `compute_target` here?

File: examples/components/CCFRAUD/traininsilo/conda.yaml [hunk 12]
```diff
@@ -0,0 +1,15 @@
+name: mnist_train_conda_env
```
Comment (garg-amit) [inline examples/components/CCFRAUD/traininsilo/conda.yaml:1]: Rename this too

File: examples/components/CCFRAUD/traininsilo/config.yaml [hunk 13]
```diff
@@ -0,0 +1,77 @@
+
```
Comment (garg-amit) [inline examples/components/CCFRAUD/traininsilo/config.yaml:1]: this file is not required, right?

File: examples/components/CCFRAUD/upload_data/conda.yaml [hunk 14]
```diff
@@ -0,0 +1,13 @@
+name: mnist_train_conda_env
```
Comment (garg-amit) [inline examples/components/CCFRAUD/upload_data/conda.yaml:1]: this too

File: examples/pipelines/ccfraud/submit.py [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+
+ML_CLIENT = connect_to_aml()
+
+####################################
+### LOAD THE PIPELINE COMPONENTS ###
+####################################
+
+# Loading the component from their yaml specifications
+preprocessing_component = load_component(
+    source=os.path.join(COMPONENTS_FOLDER, "preprocessing", "preprocessing.yaml")
```
Comment (garg-amit) [inline examples/pipelines/ccfraud/submit.py:128]: let's change this to `spec.yaml`

File: examples/pipelines/ccfraud/submit.py [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+)
+
+training_component = load_component(
+    source=os.path.join(COMPONENTS_FOLDER, "traininsilo", "config.yaml")
```
Comment (garg-amit) [inline examples/pipelines/ccfraud/submit.py:132]: this as well

**Review by garg-amit — COMMENTED** [review]
let's ensure that we have a docstring for each fn explaining all parameters

File: examples/components/utils/aggregatemodelweights/run.py [hunk 17]
```diff
@@ -65,7 +65,9 @@ def add_model(self, model_path: str):
         """
         if self.avg_state_dict is None:
             # no model yet, nothing to average
-            self.avg_state_dict = torch.load(model_path)
+            self.avg_state_dict = torch.load(
+                model_path, map_location=torch.device("cpu")
```
Comment (thomasp-ms) [inline examples/components/utils/aggregatemodelweights/run.py:69]: Please forgive my ignorance - can you explain what this does, why it is required, and what are the implications if someone wants to run this on a gpu for some reason? 

I'm sure I'm missing something, but it does look like we're hardcoding the machine type. Hence my concern.

File: docs/real-world-examples/ccfraud.md [hunk 18]
```diff
@@ -0,0 +1,35 @@
+# Finance example for Federated Learning in Azure ML
+
+This is a short example where we showcase possibilities of using Azure Machine Learning(AML) for training a model for credit card fraud detection in federating learning fashion while using variety of models. This is common scenario for credit card companies, banks and othe similar financial institutions.
```
Comment (thomasp-ms) [inline docs/real-world-examples/ccfraud.md:3]: First sentence is very long and could be split in 2. Typo in second sentence ("othe").

File: docs/real-world-examples/ccfraud.md [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+
+:warning: This is an example for demonstration purposes only and the authors of this repository do not take any responsibility for loss or harm made by its usage
```
Comment (thomasp-ms) [inline docs/real-world-examples/ccfraud.md:5]: dot at the end

File: docs/real-world-examples/ccfraud.md [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+
+As for the source of the data for this example we have used Kaggle dataset, [**Credit Card Transactions Fraud Detection Dataset**](https://www.kaggle.com/datasets/kartik2112/fraud-detection?datasetId=817870&sortBy=voteCount&types=competitions) generated using simulation that contains both genuine and fraudulent transactions. Compared to other dataset this one does not only contain an abstraction of underlying features but rather all features in their raw format. This allows us to split data into distinct geo-location and also let us show how to do a simple **preprocessing** on the **tabular data**.
+
+# Getting started
+
+### Provisioning
+Please follow the instructions in the [quickstart](../quickstart.md) to provision an open sandbox. Make note of the name of the resource group you provisioned, as well as the name of the workspace.
+
+### Kaggle credentials
+In the resource group, provisioned in the next step, lookup the Key Vault and open "Secrets" tab. Here we will need to create two plian text secrets with following names and contents:
```
Comment (thomasp-ms) [inline docs/real-world-examples/ccfraud.md:20]: next -> previous?

File: examples/pipelines/utils/upload_data/submit.py [hunk 21]
```diff
@@ -0,0 +1,247 @@
...
+    help="path to a config yaml file",
+)
+parser.add_argument(
+    "--submit",
+    default=False,
+    action="store_true",
+    help="actually submits the experiment to AzureML",
+)
+
+parser.add_argument(
```
Comment (thomasp-ms) [inline examples/pipelines/utils/upload_data/submit.py:51]: minor: why even have this as an argument, since there's just one allowed choice and it is populated by default anyway?

File: examples/components/MNIST/aggregatemodelweights/run.py [hunk 22]
```diff
@@ -82,7 +84,7 @@ def add_model(self, model_path: str):
 
         else:
             # load the new model
-            model_to_add = torch.load(model_path)
+            model_to_add = torch.load(model_path, map_location=torch.device("cpu"))
```
[the remark this answers is not in the document]
Comment (majercakdavid) [reply examples/components/MNIST/aggregatemodelweights/run.py]: This is necessary if you are training on the GPU but not running aggregation on GPU

[the remark this answers is not in the document]
Comment (majercakdavid) [reply examples/components/MNIST/aggregatemodelweights/run.py]: This forcefully loads the model to CPU instead of GPU otherwise the model is always loaded based on the device where it was trained

https://stackoverflow.com/questions/65882750/please-use-torch-load-with-map-location-torch-devicecpu

**Review by thomasp-ms — CHANGES_REQUESTED** [review]
Overall, this looks correct. There is also always the opportunity to refine later.

As far as I'm concerned, the points that need addressing before this can go in are:
- a few typos in `ccfraud.md` (fairly minor)
- my question about the modification in the generic aggregation component (this one is more important, but maybe I'm worrying for nothing?)

File: examples/components/utils/aggregatemodelweights/run.py (hunk 17 above)
Comment (majercakdavid) [reply examples/components/utils/aggregatemodelweights/run.py]: By default PyTorch tries to deserialize model on the device it was trained on and thus if the model was trained on CUDA device but there is no CUDA available on the cluster used for aggregating weights, this will fail. More info here: https://stackoverflow.com/questions/67934005/pytorch-cpu-cuda-device-load-without-gpu 

I have updated it to not to be hardcoded, even though it would always work, to always used cuda if available otherwise use the cpu. I hope this is nicer solution :)

**Comment by majercakdavid** [comment]
> Overall, this looks correct. There is also always the opportunity to refine later.
> 
> As far as I'm concerned, the points that need addressing before this can go in are:
> 
> * a few typos in `ccfraud.md` (fairly minor)
> * my question about the modification in the generic aggregation component (this one is more important, but maybe I'm worrying for nothing?)

Yes, it's bit more worrying than necessary. This has to do with the fact that PyTorch uses the same device for deserialization that was used for training unless explicitly specified otherwise. Thus, forceful mapping to CPU will always work but may not be optimal, however **skipping this parameter is the most dangerous approach**. The current update always check if the CUDA is available and if it is it uses it otherwise it uses CPU.

**Comment by thomasp-ms** [comment]
> > Overall, this looks correct. There is also always the opportunity to refine later.
> > As far as I'm concerned, the points that need addressing before this can go in are:
> > 
> > * a few typos in `ccfraud.md` (fairly minor)
> > * my question about the modification in the generic aggregation component (this one is more important, but maybe I'm worrying for nothing?)
> 
> Yes, it's bit more worrying than necessary. This has to do with the fact that PyTorch uses the same device for deserialization that was used for training unless explicitly specified otherwise. Thus, forceful mapping to CPU will always work but may not be optimal, however **skipping this parameter is the most dangerous approach**. The current update always check if the CUDA is available and if it is it uses it otherwise it uses CPU.
> 
> https://stackoverflow.com/questions/67934005/pytorch-cpu-cuda-device-load-without-gpu

Isn't it possible to include code to identify which of cpu/gpu, then set it properly?

File: docs/real-world-examples/ccfraud.md [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+# Data
+
+As for the source of the data for this example we have used Kaggle dataset, [**Credit Card Transactions Fraud Detection Dataset**](https://www.kaggle.com/datasets/kartik2112/fraud-detection?datasetId=817870&sortBy=voteCount&types=competitions) generated using simulation that contains both genuine and fraudulent transactions. Compared to other dataset this one does not only contain an abstraction of underlying features but rather all features in their raw format. This allows us to split data into distinct geo-location and also let us show how to do a simple **preprocessing** on the **tabular data**.
+
+# Getting started
+
+### Provisioning
+Please follow the instructions in the [quickstart](../quickstart.md) to provision an open sandbox. Make note of the name of the resource group you provisioned, as well as the name of the workspace.
+
+### Kaggle credentials
```
[the remark this answers is not in the document]
Comment (majercakdavid) [reply docs/real-world-examples/ccfraud.md:19]: @thomasp-ms yes that's right but this should rather be changes in the bicep template than addressed here, just like we spoke with Jeff about opening the issue

**Comment by majercakdavid** [comment]
> > > Overall, this looks correct. There is also always the opportunity to refine later.
> > > As far as I'm concerned, the points that need addressing before this can go in are:
> > > 
> > > * a few typos in `ccfraud.md` (fairly minor)
> > > * my question about the modification in the generic aggregation component (this one is more important, but maybe I'm worrying for nothing?)
> > 
> > 
> > Yes, it's bit more worrying than necessary. This has to do with the fact that PyTorch uses the same device for deserialization that was used for training unless explicitly specified otherwise. Thus, forceful mapping to CPU will always work but may not be optimal, however **skipping this parameter is the most dangerous approach**. The current update always check if the CUDA is available and if it is it uses it otherwise it uses CPU.
> > https://stackoverflow.com/questions/67934005/pytorch-cpu-cuda-device-load-without-gpu
> 
> Isn't it possible to include code to identify which of cpu/gpu, then set it properly?

I doubt that as you need to specify it already during the loading - though the current solution is the best as you can decide to run aggregation on different compute than training and so even though the training was done on the GPU you are forced to enforce CPU loading

**Comment by thomasp-ms** [comment]
> o with the fact that PyTorch uses the same device for deserialization that was used for training unless explicitly specified otherwise. Thus, forceful mapping to CPU will always work but may not be optimal, however **skipping this parameter is the most dangerous approach**. The current update always check if the CUDA is available and if it is it uses it otherwise it uses CPU.

I'm not quite following what you're saying, but I see that now you're doing a `device = torch.device("cuda" if torch.cuda.is_available() else "cpu")`. All good!

**Review by thomasp-ms — APPROVED** [review]

**Review by garg-amit — APPROVED** [review]
Looks good to me overall! It's worth adding data upload instructions in the `ccfraud.md` file.

File: docs/real-world-examples/ccfraud.md [hunk 24]
```diff
@@ -0,0 +1,50 @@
+# Finance example for Federated Learning in Azure ML
+
+This is a short example where we showcase possibilities of using Azure Machine Learning(AML) for training a model for credit card fraud detection in federating learning fashion. The example utilizes multitude of model architectures to demonstrate versatility of the proposed solution. We have chosen this scenario because we believe it is very common for credit card companies, banks and other similar financial institutions.
+
+:warning: This is an example for demonstration purposes only and the authors of this repository do not take any responsibility for loss or harm made by its usage.
```
Comment (ShizeSu06) [inline docs/real-world-examples/ccfraud.md:5]: Do we want to keep this warning here? I don't have strong preference to keep this warning or not, but if we do decide to keep this warning, we might probably be consistent and keep same warning across all our 3 examples. It looks a bit weird to me that we used this warning in one example but not other 2 examples, since I see nothing special about this example? Or the warning is added simply because it is "financial" area? Thoughts @thomasp-ms @garg-amit @majercakdavid ?

File: docs/real-world-examples/ccfraud.md [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+
+### Model choice
+Please update [**config**](../../examples/pipelines/ccfraud/config.yaml), field `model_name` in the `training_parameters` section, to reflect desired model to be trained, options include: SimpleLinear, SimpleLSTM, SimpleVAE
+
+- **SimpleLinear** - model fully composed of `torch.Linear` layers with `ReLU` activations, takes data as-is sample-by-sample
+- **SimpleLSTM** - model composed by 4 LSTM layers connected to linear head with architecture similar to **SimpleLinear**, takes data ordered by time in sequences that overlap each other
+- **SimpleVAE** - model composed of 2 encoder LSTM layers and 2 decoder LSTM layers that tries to recreate consumed sequence of transactions, the latent space created by encoder is consumed by a linear layer to perform prediction, takes data ordered by time in sequences that overlap each other
+
+## Running example
+- Update [**config**](../../examples/pipelines/ccfraud/config.yaml) to reflect your AML FL orchestration setup
```
Comment (ShizeSu06) [inline docs/real-world-examples/ccfraud.md:47]: Here might be good to add a simple sentence explaining a bit more what needs to be updated/adjusted, e.g., in Amit/Thomas PRs, they added this "(if you kept everything default you'll only have to adjust subscription id, resource group, and workspace name)". Might be good to add sth similar here for easier consumption by the customers.

File: docs/real-world-examples/ccfraud.md (hunk 24 above)
Comment (garg-amit) [reply docs/real-world-examples/ccfraud.md:5]: IMO we can remove this warning as we already have this note/warning `For the sake of simplicity, we will only provision an _open_ setup. Do not upload sensitive data to it!`

**Review by ShizeSu06 — APPROVED** [review]
